src/audio/codec_adapter/codec/waves.c

```diff
@@ -0,0 +1,756 @@
+{
+	struct codec_config *cfg;
+	struct codec_data *codec = comp_get_codec(dev);
+	struct codec_param *param;
+	uint32_t index;
+	uint32_t param_number = 0;
+	int ret = 0;
+
+	comp_info(dev, "waves_codec_configure() start type %d", type);
+
+	cfg = (type == CODEC_CFG_SETUP) ? &codec->s_cfg : &codec->r_cfg;
+
+	comp_info(dev, "waves_codec_configure() config %p, size %d, avail %d",
+		  cfg->data, cfg->size, cfg->avail);
+
+	if (!cfg->avail || !cfg->size) {
+		comp_err(dev, "waves_codec_configure() no config for type %d, avail %d, size %d",
+			 type, cfg->avail, cfg->size);
+		return -EINVAL;
+	}
+
+	if (cfg->size > MAX_CONFIG_SIZE_BYTES) {
+		comp_err(dev, "waves_codec_configure() size is too big %d", cfg->size);
+		return -EINVAL;
+	}
+
+	/* incoming data in cfg->data is arranged according to struct codec_param
+	 * there migh be more than one struct codec_param inside cfg->data, glued back to back
+	 */
+	for (index = 0; index < cfg->size && (!ret); param_number++) {
+		uint32_t param_data_size;
+
+		param = (struct codec_param *)((char *)cfg->data + index);
+		param_data_size = param->size - sizeof(param->size) - sizeof(param->id);
+
+		comp_info(dev, "waves_codec_configure() param num %d id %d size %d",
+			  param_number, param->id, param->size);
+
+		switch (param->id) {
+		case PARAM_NOP:
+			comp_info(dev, "waves_codec_configure() NOP");
+			ret = 0;
+			break;
+		case PARAM_MESSAGE:
+			ret = waves_effect_message(dev, param->data, param_data_size);
+			break;
+		case PARAM_REVISION:
+			ret = waves_effect_revision(dev);
+			break;
+		default:
+			ret = -EINVAL;
+			break;
+		}
+
+		index += param->size;
+	}
+
+	if (ret)
+		comp_err(dev, "waves_codec_configure() error %d", ret);
+
+	comp_dbg(dev, "waves_codec_configure() done");
+	return ret;
+}
+
+static int waves_effect_setup_config(struct comp_dev *dev)
+{
+	/* apply setup config if present */
+	struct codec_data *codec = comp_get_codec(dev);
+	int ret;
+
+	comp_dbg(dev, "waves_effect_setup_config() start");
+
+	if (!codec->s_cfg.avail && !codec->s_cfg.size) {
+		comp_err(dev, "waves_effect_startup_config() no setup config");
+		return -EINVAL;
+	}
+
+	if (!codec->s_cfg.avail) {
+		comp_warn(dev, "waves_effect_startup_config() no new setup config, using old");
+		codec->s_cfg.avail = true;
+	}
+
+	ret = waves_effect_config(dev, CODEC_CFG_SETUP);
+	codec->s_cfg.avail = false;
+
+	comp_dbg(dev, "waves_effect_setup_config() done");
+	return ret;
+}
+
+int waves_codec_init(struct comp_dev *dev)
+{
+	struct codec_data *codec = comp_get_codec(dev);
+	struct waves_codec_data *waves_codec;
+	int ret = 0;
+
+	comp_dbg(dev, "waves_codec_init() start");
+
+	waves_codec = codec_allocate_memory(dev, sizeof(struct waves_codec_data), 16);
+	if (!waves_codec) {
+		comp_err(dev, "waves_codec_init() allocate memory for waves_codec_data");
+		ret = -ENOMEM;
+	} else {
+		memset(waves_codec, 1, sizeof(struct waves_codec_data));
+		codec->private = waves_codec;
```

Comment:
What is the purpose of this memset? Why are we setting everything to `1`?

Reply:
this is an unfortunate typo, thank you!